classes/cloner.php

```diff
@@ -0,0 +1,159 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace local_clonecategory;
+
+use context_system;
+use core\task\manager;
+use core_course_category;
+use core_course_external;
+use core_php_time_limit;
+use Exception;
+use local_clonecategory\event\course_cloned;
+use local_clonecategory\task\clone_course_task;
+
+defined('MOODLE_INTERNAL') || die();
+require_once($CFG->dirroot . '/course/externallib.php');
+
+/**
+ *  Cloner class
+ *
+ * @package    local_clonecategory
+ * @copyright  2023 Matthew Hilton <matthewhilton@catalyst-au.net>
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class cloner {
+    /**
+     * Prepares the clone by creating the categories if necessary.
+     * @param object $data form data.
+     * @return array array containg the source and destination course categories.
+     */
+    public static function prepare($data) {
+        global $DB;
+
+        // Clone the category.
+
+        // Category selection.
+        $src = core_course_category::get($data->source);
+        $dest = core_course_category::get($data->destination);
+
+        // If a destination category name was supplied, create it and update the $dest object.
+        if (!empty($data->destcategoryname) && !empty($data->destcategoryidnumber)) {
+            if ($rec = $DB->get_record('course_categories', ['name' => trim($data->destcategoryname),
+                "idnumber" => trim($data->destcategoryidnumber), "parent" => $dest->id])) {
+                // We have an existing destination with all these details, use that one.
+                $dest = core_course_category::get($rec->id);
+            } else {
+                $dest = core_course_category::create([
+                    "name" => trim($data->destcategoryname),
+                    "idnumber" => trim($data->destcategoryidnumber),
+                    "parent" => $dest->id
+                ]);
+            }
+        }
+
+        return [$src, $dest];
+    }
+
+    /**
+     * Queues the cloning of the courses in the source category.
+     * @param object $data form data
+     * @param core_course_category $src source category
+     * @param core_course_category $dest destination category
+     */
+    public static function queue($data, core_course_category $src, core_course_category $dest) {
+        $courseids = array_keys($src->get_courses(['recursive' => false]));
+
+        foreach ($courseids as $courseid) {
+            $task = new clone_course_task();
+            $task->set_custom_data((object) [
+                'courseid' => $courseid,
+                'destid' => $dest->id,
+                'srcid' => $src->id,
+                'data' => $data
```

Comment:
do we need $data here? In theory all you need is the from, to and course id to uniquely identify this. Otherwise there might be junk in $data which makes it no longer de-dup if run again

Reply:
if there is other data, eg dates then perhaps make them explicit top level top level values instead of nested inside $data

Reply:
I didn't look too much into it first since I just left the logic as is. But it seems its just start date and enddate, so I just broke those out into top level params